### DISCUSSION TITLE
How to change application entry point to a custom address?

### DISCUSSION THREAD

**Discussion by shalauko** [zephyrproject-rtos/zephyr#79763] [Q&A]
I'm trying to use "settings" subsystem to store some variables between device boots. But failed to implement correct DTS.

I work on a project where we will have a custom board, therefore I want to have a custom `.dts `file. For development purpose we use on STM32 NUCLEO-F446RE. The target board will be based on STM32F446RE, so I took the [available](https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/st/nucleo_f446re/nucleo_f446re.dts) `.dts` file for NUCLEO-F446RE as a base.

The mentioned devicetree contains flash section adjusted to use with bootloader. I.e. it has first `boot_partition`, then `storage_partition` and then few sections of image. As long as firmware was less than 64KB it worked fine without bootloader. So, basically we put firmware in the bootloader section, but it was fine. However, our firmware is bigger that 64KB, so we needed to move it to a first image section. Therefore, we needed to implement simple bootloader which only jump to `slot0_partition` and nothing more, when real application is flashed in `slot0_partition`.

Our current DTS looks in the following way and it works fine:
```
&flash0 {
	partitions {
		compatible = "fixed-partitions";
		#address-cells = <1>;
		#size-cells = <1>;

		bootloader: partition@0 {
			label = "bootloader";
			reg = <0x0 DT_SIZE_K(16)>;
			read-only;
		};

		storage_partition: partition@4000 {
			label = "storage";
			reg = <0x4000 DT_SIZE_K(48)>;
		};

		slot0_partition: partition@10000 {
			label = "image-0";
			reg = <0x00010000 DT_SIZE_K(448)>;
		};
	};
};
```
I want to simplify the whole process as we don't actually need a bootloader. Actually, we need one application section and one "storage" section.

I've tried to follow an example from [Zephyr documentation](https://docs.zephyrproject.org/latest/develop/flash_debug/host-tools.html) (with a ROM bootloader), but then `settings_backend_init` fails with code `-EDOM` (-33) in this part of code (my nvs_sector_size at the end of the flash is 128KB):
```
<snip>
	nvs_sector_size = CONFIG_SETTINGS_NVS_SECTOR_SIZE_MULT *
			  hw_flash_sector.fs_size;

	if (nvs_sector_size > UINT16_MAX) {
		return -EDOM;
	}
<snip>
```

Then I changed order of sections, but in this case the board doesn't boot at all.
```
/ {
	chosen {
		zephyr,code-partition = &code_partition;
	};
};

&flash0 {
        partitions {
                compatible = "fixed-partitions";
                #address-cells = <1>;
                #size-cells = <1>;

			storage_partition: partition@0 {
                        label = "storage";
                        reg = <0x00000000 0x10000>;
                };

                code_partition: partition@10000 {
                        label = "code";
                        reg = <0x10000 0x70000>;
                        read-only;
                };
        };
};
```

Could you please advise how should configure my `.dts` and `prj.conf` to have storage at the beginning of the flash and entrance point moved 64KB further? Technically speaking, there is no requirement to have storage at the beginning of the flash. The one requirement is to get rid of bootloader and keep settings subsystem operable.

**Comment by Laczen** [comment]
Hi @shalauko some devices have the possibility to change the bootaddress in the option bytes. Sadly your device does not have this property.

The only way you can reserve sectors for nvs on your device is by using a bootloader (this can be a very simple one) on the first sector of your device. All this bootloader needs to do is jump to a different address (the slot0 address). You will have to create this simple loader (which can be a zephyr app) yourself and program it to the first sector of the device.

**Reply by enthunilu** [reply]
I think available MCUBoot can be used as a bootloader and as per the default memory partition
[GitLink](https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/st/nucleo_f446re/nucleo_f446re.dts) you can decide on where to jump by using

```
	chosen {
		zephyr,code-partition = &slot0_partition;
	};
```
